Rust/src/sampled_tree.rs

```diff
@@ -0,0 +1,267 @@
+    /// assert_eq!(tree.num_observations(), 2);
+    /// ```
+    pub fn update(&mut self, point: Vec<T>, sequence_index: usize) {
+        // we need a point key that we can submit to the sampler. the strategy,
+        // then, is to first add the point to the tree and then sample using
+        // the output key. if the key is accepted by the sampler then we
+        // evict if necessary. otherwise, we delete the point we just added
+        //
+        // TODO: is there a way to do this without cloning? We need a reference
+        // to the point if we need to delete afterward. Slabs allow you to
+        // obtain the next available key...
+        let point_key = match self.tree.add_point(point.clone()) {
+            AddResult::AddedPoint(key) => key,
+            AddResult::MassIncreased(key) => key,
+        };
+
+        match self.sampler.sample(point_key, sequence_index) {
+            SamplerResult::Accepted(evicted) => match evicted {
+                Some(evicted) => {
+                    // TODO: can we satisfy the borrow checker so that we can
+                    // perform the delete without needing to clone the point?
+                    let evicted_point = {
+                        let point_store = self.point_store.borrow();
+                        point_store.get(*evicted.value()).unwrap().clone()
+                    };
+                    self.tree.delete_point(&evicted_point);
+                }
+                None => ()
+            },
+            SamplerResult::Ignored => { self.tree.delete_point(&point); }
+        }
+    }
+
+    /// Traverse the tree with a given query point as input.
+    ///
+    /// Returns an iterator on the nodes of the tree. The iterator begins at the
+    /// tree's root node and follows the branch to the leaf node containing the
+    /// point nearest to the query point in L1 norm.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::{Node, SampledTree};
+    /// let mut tree: SampledTree<f32> = SampledTree::new(32, 0.01);
+    /// tree.update(vec![0.0, 0.0], 0);
+    /// tree.update(vec![1.0, 1.0], 1);
+    ///
+    /// // provide a query point that will always be to the left of any cut
+    /// // on the two points inserted above, no matter the cut location
+    /// let query = vec![-2.0, -2.0];
+    /// let nodes: Vec<&Node<f32>> = tree.traverse(&query).collect();
+    /// assert_eq!(nodes.len(), 2);
+    ///
+    /// // verify the last point of the tree is equal to the left-most point
+    /// if let Node::Leaf(leaf) = nodes[1] {
+    ///     let point_key = leaf.point();
+    ///     let point_store = tree.borrow_point_store();
+    ///     let point = point_store.get(point_key).unwrap();
+    ///     assert_eq!(point, &vec![0.0, 0.0]);
+    /// } else {
+    ///     panic!("Last node in traversal should be a leaf!")
+    /// }
+    /// ```
+    ///
+    pub fn traverse<'a>(&'a self, point: &'a Vec<T>) -> NodeTraverser<'a, T> {
```

Comment:
What is the use case? What types, other than `Tree`, would we want to traverse?

I suppose you can make an argument that since both `Tree` and `SampledTree` are exposing a `traverse()` then that could justify the creation of a trait. But the `SampledTree::traverse()` is just a convenience method.

Reply:
That's a good question. We initially introduced the "traversable" interface because a forest or group of trees could also be traversable, and we wanted to enable distributed forests whose components were sub-forests. That said, we didn't close on a design that we liked for this piece, so currently the only things that implement traversable are tree and sampler-plus-tree. When deciding what things to define as traits in the rust library, we should be thinking about which pieces we might want to support different implementations for (with different performance characteristics or data layouts).

Reply:
> When deciding what things to define as traits in the rust library, we should be thinking about which pieces we might want to support different implementations for (with different performance characteristics or data layouts).

Totally understand. The visitors are a clear example of this strategy design pattern. (I'm predicting that I'll need to adopt the trait approach two PRs from now.) I just want to balance wanting to make all of these relationships explicit with practicality. It's easy for me to implement a `Traversable` trait but I also don't want to add some feature that isn't (yet) necessary.

Still, put this down as Issue #186 and will address it later.